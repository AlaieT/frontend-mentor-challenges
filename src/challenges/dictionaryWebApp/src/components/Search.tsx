import React from "react";

import classNames from "../utils/classNames";

import type { SearchProps } from "../types";

import { ReactComponent as SearchIcon } from "../assets/images/icon-search.svg";

import styles from "../styles/components/search.module.scss";

const Search = ({
  value,
  mode,
  url,
  callback = () => undefined
}: SearchProps) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [isError, setError] = React.useState(false);
  const [isSearch, setSearch] = React.useState(false);

  const handleSubmit = React.useCallback(
    (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();

      if (inputRef.current?.value) {
        if (isError) setError(false);
        if (window) window.scrollTo({ top: 0, behavior: "auto" });

        setSearch(true);

        fetch(url + inputRef.current.value, { method: "GET" })
          .then((response) => response.json())
          .then((result) => {
            setSearch(false);
            callback(result);
          })
          .catch(() => {
            setSearch(false);
            callback(null);
          });
      } else {
        setError(true);
        callback(null);
      }
    },
    [url, isError, callback]
  );

  React.useEffect(() => {
    if (value && inputRef.current) {
      if (window) window.scrollTo({ top: 0, behavior: "auto" });

      inputRef.current.value = value;

      setSearch(true);

      fetch(url + value, { method: "GET" })
        .then((response) => response.json())
        .then((result) => {
          setSearch(false);
          callback(result);
        })
        .catch(() => {
          setSearch(false);
          callback(null);
        });
    }
  }, [value, url, callback]);

  return (
    <form id={styles["search-form"]} onSubmit={handleSubmit}>
      <div
        id={styles.search}
        className={classNames({
          [styles[mode]]: mode === "dark",
          [styles.error]: isError,
          [styles.search]: isSearch
        })}
      >
        <input
          ref={inputRef}
          id="search-input"
          type="text"
          placeholder="Search for any word…"
          disabled={isSearch}
        />
        <button type="submit">
          <SearchIcon width={15.55} height={15.55} />
        </button>
      </div>
      {isError && (
        <label id={styles["label-error"]} htmlFor="search-input">
          Whoops, can’t be empty…
        </label>
      )}
    </form>
  );
};

export default Search;
