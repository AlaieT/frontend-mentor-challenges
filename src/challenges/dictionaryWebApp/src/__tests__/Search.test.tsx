import React from "react";
import {
  render,
  cleanup,
  screen,
  act,
  fireEvent
} from "@testing-library/react";

import Search from "../components/Search";

afterEach(() => cleanup());

describe("Search", () => {
  describe("correct render", () => {
    it("should match snapshot", () => {
      expect(
        render(<Search mode="light" url="/word" />).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should give an error when search word is empty string", async () => {
      render(<Search mode="light" url="/word" />);

      fireEvent.input(screen.getByRole("textbox"), { target: { value: "" } });

      await act(() => fireEvent.click(screen.getByRole("button")));

      expect(
        screen.getByLabelText("Whoops, can’t be empty…")
      ).toBeInTheDocument();
    });

    it("should fetch word definition onsubmit", async () => {
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () => Promise.resolve({ word: "response" })
        })
      );

      render(<Search mode="light" url="/word/response" />);

      fireEvent.input(screen.getByRole("textbox"), {
        target: { value: "response" }
      });

      await act(() => fireEvent.click(screen.getByRole("button")));

      expect(global.fetch).toBeCalled();
    });
  });
});
