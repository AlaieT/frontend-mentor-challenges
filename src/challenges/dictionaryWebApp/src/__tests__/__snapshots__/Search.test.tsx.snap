// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search correct render should match snapshot 1`] = `
<DocumentFragment>
  <form
    id="search-form"
  >
    <div
      id="search"
    >
      <input
        id="search-input"
        placeholder="Search for any wordâ€¦"
        type="text"
      />
      <button
        type="submit"
      >
        <div
          height="15.55"
          width="15.55"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
